source/code/include/playfab/PlayFabPlatformUtils.h

```diff
@@ -7,53 +7,127 @@
-        gmtime_s(&timeInfo, &now);
-#elif defined(PLAYFAB_PLATFORM_LINUX) || defined(PLAYFAB_PLATFORM_IOS) || defined(PLAYFAB_PLATFORM_ANDROID) || defined(PLAYFAB_PLATFORM_PLAYSTATION)
-        timeInfo = *gmtime(&now);
+        tm timeInfo{ 0 };
+#if defined(PLAYFAB_PLATFORM_PLAYSTATION)
+        gmtime_s(&input, &timeInfo);
+#elif defined(PLAYFAB_PLATFORM_WINDOWS) || defined(PLAYFAB_PLATFORM_XBOX)
+        gmtime_s(&timeInfo, &input);
+#else
+        gmtime_r(&input, &timeInfo);
 #endif
-        char buff[64];
-        strftime(buff, 64, "%Y-%m-%dT%H:%M:%S.000Z", &timeInfo);
+        return timeInfo;
+    }
 
-        return buff;
+    inline time_t UtcTmToTimeT(tm input)
+    {
+#if defined(PLAYFAB_PLATFORM_PLAYSTATION)
+        return mktime(&input);
+#elif defined(PLAYFAB_PLATFORM_WINDOWS) || defined(PLAYFAB_PLATFORM_XBOX)
+        return _mkgmtime(&input);
+#else
+        return timegm(&input);
+#endif
+    }
+
+    inline tm TimePointToUtcTm(TimePoint input)
+    {
+        return TimeTToUtcTm(Clock::to_time_t(input));
     }
 
-    inline std::string LocalTimePointToUtcString(TimePoint now)
+    inline TimePoint UtcTmToTimePoint(const tm& input)
     {
-        return LocalTimeTToUtcString(std::chrono::system_clock::to_time_t(now));
+        return TimeTToTimePoint(UtcTmToTimeT(input));
     }
 
-    inline time_t UtcStringToLocalTimeT(const std::string& utcString)
+    // Get Time now - Platform dependent granularity (granularity: upto 1 second, accuracy within a few seconds)
+    inline TimePoint GetTimePointNow()
     {
-        time_t output;
-        tm timeStruct = {};
+        // The conversion is mostly to ensure consistent behavior among all platforms
+        return std::chrono::time_point_cast<std::chrono::seconds>(Clock::now());
+    }
 
-        std::istringstream iss(utcString);
-        iss >> std::get_time(&timeStruct, "%Y-%m-%dT%T");
-        timeStruct.tm_isdst = 0;  // 0 means "not in DST" PlayFab assumes UTC/Zulu always
-#if defined(PLAYFAB_PLATFORM_PLAYSTATION)
-        output = mktime(&timeStruct);
-#elif defined(PLAYFAB_PLATFORM_IOS) || defined(PLAYFAB_PLATFORM_ANDROID) || defined(PLAYFAB_PLATFORM_LINUX)
-        gmtime_r(&output, &timeStruct);
+    inline time_t GetTimeTNow()
+    {
+        return TimePointToTimeT(GetTimePointNow());
+    }
+
+    // Get a tick count that represents now in milliseconds (not useful for absolute time)
+    inline long GetMilliTicks()
+    {
+#if defined(PLAYFAB_PLATFORM_SWITCH)
+        static_assert("You must request the Nintendo specific XPlat SDK from PlayFab support.");
 #else
-        output = _mkgmtime(&timeStruct);
+        auto msClock = std::chrono::time_point_cast<std::chrono::milliseconds>(Clock::now());
+        return static_cast<long>(msClock.time_since_epoch().count());
 #endif
+    }
+
+    // Time Serialization
+    inline std::string UtcTmToIso8601String(const tm& input)
+    {
+        char buff[64];
```

Comment:
magic numbers should be avoided. is there anyway you can tie this back to the WRITE_FORMAT?
e.g. char buff[] = "<an example timestring which is the max size that TIEMSTAMP_WRITE_FORMAT could write>";

Reply:
64 was chosen as "the first bigger 2^n number after the documented max number (~36 maybe)" for one of the methods used in the previous implementation (though I don't know which off-hand).
It's also similarly well-larger than the biggest string PlayFab is going to send.
It's always going to be fairly arbitrary, because the C# impl in PlayFab is not strictly sized like C++ is, so we just want to pick a number that's sufficiently flexible for what C# might do.
I do agree it could be a constant though.

---
In reply to: [357802305](https://github.com/PlayFab/XPlatCoreTemplate/pull/65#discussion_r357802305) [](ancestors = 357802305)